modules/programs/brew-bundle.nix

```diff
@@ -0,0 +1,225 @@
+
+        When this option is used, <literal>"mas"</literal> is automatically added to
+        <option>programs.brew-bundle.brews</option>.
+
+        Note that you need to be signed into the Mac App Store for <command>mas</command> to
+        successfully install and upgrade applications, and that unfortunately apps removed from this
+        option will not be uninstalled automatically even if
+        <option>programs.brew-bundle.cleanup</option> is set to <literal>"uninstall"</literal>
+        or <literal>"zap"</literal> (this is currently a limitation of Homebrew Bundle).
+
+        For more information on <command>mas</command> see: https://github.com/mas-cli/mas
+      '';
+    };
+
+    whalebrews = mkOption {
+      type = with types; listOf str;
+      default = [];
+      example = [ "whalebrew/wget" ];
+      description = ''
+        Docker images to install using <command>whalebrew</command>.
+
+        When this option is used, <literal>"whalebrew"</literal> is automatically added to
+        <option>programs.brew-bundle.brews</option>.
+
+        For more information on <command>whalebrew</command> see:
+        https://github.com/whalebrew/whalebrew
+      '';
+    };
+
+    extraConfig = mkOption {
+      type = types.lines;
+      default = "";
+      example = ''
+        # 'brew tap' with custom Git URL
+        tap "user/tap-repo", "https://user@bitbucket.org/user/homebrew-tap-repo.git"
+
+        # set arguments for all 'brew cask install' commands
+        cask_args appdir: "~/Applications", require_sha: true
+
+        # 'brew install --with-rmtp', 'brew services restart' on version changes
+        brew "denji/nginx/nginx-full", args: ["with-rmtp"], restart_service: :changed
+        # 'brew install', always 'brew services restart', 'brew link', 'brew unlink mysql' (if it is installed)
+        brew "mysql@5.6", restart_service: true, link: true, conflicts_with: ["mysql"]
+
+        # 'brew cask install --appdir=~/my-apps/Applications'
+        cask "firefox", args: { appdir: "~/my-apps/Applications" }
+        # 'brew cask install' only if '/usr/libexec/java_home --failfast' fails
+        cask "java" unless system "/usr/libexec/java_home --failfast"
+      '';
+      description = "Extra lines to be added verbatim to the generated Brewfile.";
+    };
+  };
+
+  config = {
+    assertions = mkIf cfg.enable [
+      {
+        assertion = builtins.pathExists /usr/local/bin/brew;
+        message = ''
+          Homebrew not installed.
+
+          Please install Homebrew yourself before using the programs.brew-bundle module.
+
+          See installation instructions at: https://brew.sh
+        '';
+      }
+    ];
+
+    programs.brew-bundle.brews =
+      optional (cfg.masApps != {}) "mas" ++
+      optional (cfg.whalebrews != []) "whalebrew";
+
+    environment.variables = mkIf cfg.enable (
+      (if cfg.userConfig.brewfile then { HOMEBREW_BUNDLE_FILE = "${brewfile}"; } else {}) //
```

Comment:
Is there a reason to expose these? Looks kind of redundant since activation already applies the brewfile and the file itself is readonly.

Reply:
TLDR: I think it's handy that this module sets `HOMEBREW_BUNDLE_FILE` for me in my environment and I expect some others to also appreciate it. I'd be happy in a world where the module just set this variable in my environment when enabled, and didn't have an option, but I'm worried there may be people that have use-cases where not being able to disable setting that variable in their environment would be annoying. 

---

I like having the `HOMEBREW_BUNDLE_FILE` variable set in my environment, since sometimes I do still want to run `brew bundle` myself. 

For example, I sometimes install some packages/apps (that aren't available through Nix) using Homebrew manually to try them out, or because I only need them temporarily. With this environment variable set, I can run `brew bundle cleanup` to get a list of which things are installed on my system using Homebrew that aren't in my Brewfile. I might also want to run `brew bundle cleanup --force` to remove them.

Since I set `homebrew.cleanup` to `"zap"`, I could instead just run `darwin-rebuild switch`, but sometimes I don't want to run a full build/activation just for this. Also, I can imagine some people wanting to do similar things, but who prefer to have `homebrew.cleanup` set to `"none"`, and so the only way to do a cleanup would be to run the command manually. Without that variable set, one would have to manually specify the path to the generated Brewfile in the store which would be a pain.

The above is an example of why it's handy for this module to set `HOMEBREW_BUNDLE_FILE` in the users environment, but it isn't necessarily and argument that there should be an option for it. In my case, I basically never expect to set `homebrew.userConfig.brewfile` to `false`, so I wouldn't mine the module simply adding it to `environment.variables` when `homebrew.enable` is set to `true`. The reason I decided to included it as an option is that, I have a vague concern that it might be annoying for some people, though I don't have a great sense of why it might be. Maybe it's not uncommon for people to use multiple Brewfiles for some reason? ¯\_(ツ)_/¯

For completeness, I should acknowledge that this module doesn't *need* to have the ability to set `HOMEBREW_BUNDLE_FILE` at all. Users like me who care about it could just set it manually ourselves. My take is that setting it (and maybe also including an option to disable setting it) matches what users expect of modules like this. They way I see it, this module is about declaratively managing Homebrew Bundle. From that lens, it would seem wrong to me if, by default when I enabled this module, the `brew bundle` command didn't know about the Brewfile this module generated.

Reply:
This sounds like a very specific workflow to me, which maybe shouldn't be the default. But I have no idea 😄 .

Reply:
Alrighty, I'm going to go with removing the options, and setting both variables when the module is enabled.